rfc/2020/conformance-validation/mayankshah1607/rfc.md

```diff
@@ -0,0 +1,449 @@
+- [grpc/grpc-go](https://github.com/grpc/grpc-go)
+- [net/http](https://golang.org/pkg/net/http/)
+- [googollee/go-socket](https://github.com/googollee/go-socket.io)
+- [go-redis/redis](https://github.com/go-redis/redis)
+- [go-sql-driver/mysql](https://github.com/go-sql-driver/mysql)
+- [go-yaml/yaml](https://github.com/go-yaml/yaml)
+
+## Use Cases
+
+This project introduces an extensive e2e test suite that validates a cluster by interacting with various features of Linkerd, mainly those that involve communicating with a data plane. The e2e tests for conformance shall carry out checks for the following: 
+1. Automatic proxy injection
+2. linkerd tap, routes, edges, stat cmds
+3. Distributed tracing
+4. Canary release
+5. Ingress
+6. gRPC streaming (client/server/bi-directional streaming and unary)
+7. Websockets
+8. MySQL
+9. Redis
+
+Before the test suite is run, the following actions must be carried out in sequence:
+1. Check if kubectl binary is available
+2. Check if linkerd binary is available
+3. Check if connection to cluster can be established via kubectl
+4. Check if linkerd control plane is installed
+5. Install the sample application
+
+Once the test suite is complete (success / failure ) all resources created for the test are destroyed if not immediately after each test.
+
+
+## Goals
+The goal of this project is to :
+1. develop a sample application (that comprises microservices) that can exercise various features of Linkerd, mainly that involve the data plane. (from https://github.com/linkerd/linkerd2/issues/2079) This application shall also exercise various protocols and topologies that could benefit from deeper testing and surface issues around streaming, websockets, MySQL and Redis.
+2. provide an extensive e2e test suite that can make use of this application and validate many properties of the cluster, to help a user test for conformance w.r.t Linkerd, i.e to check if the cluster has been configured to work correctly over a longer period of time.
+
+## Non-goals
+It is a non-goal for this project to provide an application that is expected to be a part of Linkerd application architecture (control / data plane). This project in no way shall directly affect the way Linkerd functions. Unlike the existing integration tests, the e2e tests for conformance in this project are not expected to be a part of the Linkerd CI workflow either. These tests shall run as a stand-alone component such that it can interact with a k8s cluster.
+
+## Deliverables
+
+### Must have (shall be completed by end of GSoC):
+- An e2e test suite that can perform conformance validation for the following features:
+    1. Automatic proxy injection
+    2. Linkerd tap, stat, routes, edges cmd
+    3. Distributed tracing
+    4. Canary release
+    5. Ingress
+    6. gRPC streaming (client/server/bi-directional streaming and unary)
+    7. Websockets
+    8. MySQL
+    9. Redis
```

Comment:
There are some items in here that I feel it's less important (either because they are covered by the integration test or they don't tell us much), and those under "Good-to-have" which I think are crucial. Here's my wish list of test cases:

1. Proxy injection (automatic or not, via the webhook)
1. If service account token is missing, proxy injection should fail
1. Linkerd `tap`, `stat` (latency), `routes` (service profile), `edges` (mTLS) commands
1. `tap` extension API server is functional (`linkerd check --pre` can do that)
1. Retries and timeouts (take a look at `test/serviceprofiles/service_profiles.go` to see if you can reuse code)
1. Ingress traffic from ingress controller to the pod
1. Data plane proxies are healthy (start with `linkerd -n <ns> check --proxy`, then drill down into 503 errors in logs)
1. User's existing PodSecurityPolicy doesn't fail `kubectl apply` because the proxy sidecar doesn't have the required security context

Other less important items:

1. MySQL, Redis where you end up having to skip some ports anyway. It will still be important to know that these databases still work once their ports are skipped
1. gRPC streaming, websocket shouldn't make a difference

There are already some integration tests for tracing. Since canary relies on a number of external components, I'm concerned that the tests might be flaky. It feels more efficient to perform canary testing manually, than to write code with assumptions on external components' behaviour.